### PR TITLE
[Speculative Decoding] [BugFix] overlap compute logprobs for speculative decoding

### PR DESCRIPTION
**Pull request by huicongyao**
## Motivation

> :bulb: If this PR is a Cherry Pick, the PR title needs to follow the format by adding the [Cherry-Pick] label at the very beginning and appending the original PR ID at the end. For example, [Cherry-Pick][CI] Add check trigger and logic(#5191)

> :bulb: 如若此PR是Cherry Pick，PR标题需遵循格式，在最开始加上[Cherry-Pick]标签，以及最后面加上原PR ID，例如[Cherry-Pick][CI] Add check trigger and logic(#5191)

## Modifications

1. [Refactor] Refactor logprob computation for async scheduling support

Add CUDA kernel build_sampling_params_logprob to replace repeat_interleave for sampling parameter expansion, enabling asynchronous execution.
Refactor SpeculativeSampler.compute_logprobs to use fixed-shape padded logits, laying the groundwork for async scheduling.
Switch temp_scaled_logprobs / top_p_normalized_logprobs from static defaults to per-step dynamic evaluation based on active requests, eliminating redundant computation when no logprob requests exist.
Temporarily disable top_p_normalized_logprobs path (to be restored in a follow-up).

2. [BugFix] Fix shape mismatch when CUDA Graph is disabled

Replace hardcoded real_bsz * (max_draft_token_num + 1) with target_hidden_states.shape[0] in MTP token count calculation.
remote skip_save_output in `gpu_model_runner.py` cause it's not used

3. Fix typo specualate → speculate

Correct function names and references in GPU/XPU pre/post-process modules (post_process_speculate, save_output_speculate, xpu_post_process_speculate).

## Checklist

- [x] Add at least a tag in the PR title.
  - Tag list: [`[FDConfig]`,`[APIServer]`,`[Engine]`, `[Scheduler]`, `[PD Disaggregation]`, `[Executor]`, `[Graph Optimization]`, `[Speculative Decoding]`, `[RL]`, `[Models]`, `[Quantization]`, `[Loader]`, `[OP]`, `[KVCache]`, `[DataProcessor]`, `[BugFix]`, `[Docs]`, `[CI]`, `[Optimization]`, `[Feature]`, `[Benchmark]`, `[Others]`, `[XPU]`, `[HPU]`, `[GCU]`, `[DCU]`, `[Iluvatar]`, `[Metax]`]
  - You can add new tags based on the PR content, but the semantics must be clear.
- [x] Format your code, run `pre-commit` before commit.
- [x] Add unit tests. Please write the reason in this PR if no unit tests.
- [ ] Provide accuracy results.
- [ ] If the current PR is submitting to the `release` branch, make sure the PR has been submitted to the `develop` branch, then cherry-pick it to the `release` branch with the `[Cherry-Pick]` PR tag.

### CODE REVIEW

File: fastdeploy/worker/gpu_model_runner.py [hunk 1]
```diff
@@ -144,8 +144,8 @@ def __init__(
                 if fd_config.model_config.max_logprobs == -1
                 else fd_config.model_config.max_logprobs
             )
-        self.temp_scaled_logprobs = True
-        self.top_p_normalized_logprobs = True
+        self.temp_scaled_logprobs = False
```
[the remark this answers is not in the document]
Comment (huicongyao) [reply fastdeploy/worker/gpu_model_runner.py:148]: top_p_normalized_logprobs不应被默认启用

File: fastdeploy/model_executor/layers/sample/sampler.py [hunk 2]
```diff
@@ -684,50 +686,50 @@ def compute_logprobs(
...
-            if top_p_token_mask.any():
-                probs = F.softmax(last_logits, axis=-1)
-                probs = top_p_normalize_probs_paddle(probs, real_token_top_p)
-                top_p_logprob = paddle.log(probs)
-        if top_p_logprob is not None:
-            last_logprobs = paddle.where(top_p_token_mask, top_p_logprob, last_logprobs)
-        return last_logprobs
+            padded_logits = padded_logits / temp_temperature
+
+        last_logprobs = F.log_softmax(padded_logits, axis=-1)
```
[the remark this answers is not in the document]
Comment (huicongyao) [reply fastdeploy/model_executor/layers/sample/sampler.py:710]: 这个功能从添加后从未使用过，但是对应的逻辑会影响overlap，先注释掉，后续有需求时添加回来

File: fastdeploy/model_executor/layers/sample/sampler.py [hunk 3]
```diff
@@ -684,50 +686,50 @@ def compute_logprobs(
...
+        #     top_p_token_mask = paddle.logical_and(top_p_normalized_logprobs, real_token_top_p != 1.0)
+
+        #     probs = F.softmax(padded_logits, axis=-1)
+        #     probs = top_p_normalize_probs_paddle(probs, real_token_top_p)
+        #     top_p_logprob = paddle.log(probs)
+        # if top_p_logprob is not None:
+        #     last_logprobs = paddle.where(top_p_token_mask, top_p_logprob, last_logprobs)
+
+        # NOTE(huicongyao) temporarily used for slice last_logprobs to its real shape, remove in the future
+        real_token_num = batch_token_num.sum().item()
```
[the remark this answers is not in the document]
Comment (huicongyao) [reply fastdeploy/model_executor/layers/sample/sampler.py:733]: 后续整个链路都改为bsz * (spec step + 1)后，这里的求和&切分会去除

[the remark this answers is not in the document]
Comment (huicongyao) [reply fastdeploy/model_executor/layers/sample/sampler.py:733]: 同上，后续整个链路都改为bsz * (spec step + 1)后，这里的求和&切分会去除

File: fastdeploy/model_executor/layers/sample/sampler.py [hunk 4]
```diff
@@ -684,50 +686,50 @@ def compute_logprobs(
         share_inputs = sampling_metadata.share_inputs
         last_logits = logits
         real_bsz = share_inputs["seq_lens_this_time"].shape[0]
+
+        # NOTE(huicongyao): temporarily used to provide a max_sized input, remove in the future
+        num_tokens = real_bsz * (self.num_speculative_tokens + 1)
```
Comment (freeliuzc) [inline fastdeploy/model_executor/layers/sample/sampler.py:691]: 这里是否可以用overlap 的那个 real_bsz? seq_lens 的real_bsz 可能会大非常多

File: fastdeploy/model_executor/layers/sample/sampler.py [hunk 5]
```diff
@@ -684,50 +686,50 @@ def compute_logprobs(
...
-                probs = F.softmax(last_logits, axis=-1)
-                probs = top_p_normalize_probs_paddle(probs, real_token_top_p)
-                top_p_logprob = paddle.log(probs)
-        if top_p_logprob is not None:
-            last_logprobs = paddle.where(top_p_token_mask, top_p_logprob, last_logprobs)
-        return last_logprobs
+            padded_logits = padded_logits / temp_temperature
+
+        last_logprobs = F.log_softmax(padded_logits, axis=-1)
+        # top_p_logprob = None
```
Comment (freeliuzc) [inline fastdeploy/model_executor/layers/sample/sampler.py]: 这里注释掉的原因以及背景表述清楚

File: fastdeploy/model_executor/layers/sample/sampler.py (hunk 3 above)
Comment (freeliuzc) [inline fastdeploy/model_executor/layers/sample/sampler.py:733]: 这里是同步操作，.item 是 D->H 拷贝

File: fastdeploy/spec_decode/mtp.py [hunk 6]
```diff
@@ -894,7 +894,7 @@ def _propose_cuda(self, step_use_cudagraph: bool = False, is_dummy_run: bool = F
                 token_num_cpu = self.model_inputs["seq_lens_this_time"].numpy().sum().item()
             else:
                 if substep == 0:
-                    token_num_cpu = real_bsz * (self.max_draft_token_num + 1)
+                    token_num_cpu = self.model_inputs["target_hidden_states"].shape[0]
```
Comment (freeliuzc) [inline fastdeploy/spec_decode/mtp.py:897]: 这里为什么更改呢？感觉会有风险

File: fastdeploy/worker/gpu_model_runner.py [hunk 7]
```diff
@@ -1202,6 +1202,12 @@ def _prepare_inputs(self, cached_token_num=-1, cached_real_bsz=-1, is_dummy_or_p
             for req in self.forward_batch_reqs_list
             if req is not None and req.sampling_params is not None and req.sampling_params.logprobs is not None
         ]
+        self.temp_scaled_logprobs = bool(logprobs_reqs) and any(
```
Comment (freeliuzc) [inline fastdeploy/worker/gpu_model_runner.py]: 这里为什么用 bool 判断 dict呢

File: fastdeploy/spec_decode/mtp.py (hunk 6 above)
Comment (huicongyao) [reply fastdeploy/spec_decode/mtp.py:897]: 这是之前的一个bug，PD分离下MTP D节点关cuda graph时，mtp forward会报错shape不匹配；

File: fastdeploy/model_executor/layers/sample/sampler.py (hunk 3 above)
Comment (huicongyao) [reply fastdeploy/model_executor/layers/sample/sampler.py:733]: 后续整个链路都改为pad到最大shape后，这里的求和&切分会去除

File: fastdeploy/worker/gpu_model_runner.py (hunk 7 above)
Comment (huicongyao) [reply fastdeploy/worker/gpu_model_runner.py]: 这里logprobs_reqs是一个list，bool用来判断list是否为空

File: fastdeploy/model_executor/layers/sample/sampler.py (hunk 4 above)
Comment (huicongyao) [reply fastdeploy/model_executor/layers/sample/sampler.py:691]: 已改为使用这个real_bsz

File: tests/layers/test_sampler.py [hunk 8]
```diff
@@ -327,7 +328,7 @@ def test_speculative_sampler_basic(monkeypatch):
     m.top_p_normalized_logprobs_flag = True
     m.share_inputs = {
         "seq_lens_this_time": paddle.to_tensor([[1]], dtype="int64"),
-        "accept_num": paddle.to_tensor([1], dtype="int64"),
+        "accept_num": paddle.to_tensor([1], dtype="int32"),
     }
     gathered = sampler.gather_logprobs(sampler.compute_logprobs(logits, m), 0, paddle.to_tensor([1], dtype="int64"))
```
[the remark this answers is not in the document]
Comment (huicongyao) [reply tests/layers/test_sampler.py:333]: 这部分会进行修改

File: fastdeploy/model_executor/layers/sample/sampler.py [hunk 9]
```diff
@@ -679,55 +681,59 @@ def compute_logprobs(
...
+                top_p_normalized_logprobs[:max_occupied_slots].squeeze(1), batch_token_num, num_tokens, real_bsz
+            ).unsqueeze(1)
             top_p_token_mask = paddle.logical_and(top_p_normalized_logprobs, real_token_top_p != 1.0)
-            if top_p_token_mask.any():
-                probs = F.softmax(last_logits, axis=-1)
-                probs = top_p_normalize_probs_paddle(probs, real_token_top_p)
-                top_p_logprob = paddle.log(probs)
+
+            probs = F.softmax(padded_logits, axis=-1)
+            probs = top_p_normalize_probs_paddle(probs, real_token_top_p)
```
[the remark this answers is not in the document]
Comment (huicongyao) [reply fastdeploy/model_executor/layers/sample/sampler.py:727]: top_p_token_mask.any()的判断是冗余的，并不需要，没有logprob normalize请求时，计算top_p_token_mask的这部分代码不会执行

File: custom_ops/gpu_ops/speculate_decoding/build_sampling_params_logprob.cu [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+    const T* input_params,
+    const int32_t* token_num_per_batch,
+    const int64_t token_num_output_cpu,
+    int64_t real_bsz) {
+  const int bi = blockIdx.x;
+  const int tid = threadIdx.x;
+
+  // Compute start offset: sum of token_num_per_batch[0..bi-1]
+  int start_offset = 0;
+  for (int i = 0; i < bi; i++) {
```
[the remark this answers is not in the document]
Comment (huicongyao) [reply custom_ops/gpu_ops/speculate_decoding/build_sampling_params_logprob.cu:33]: profiler实测不会有性能问题，这部分的计算开销没那么大

File: fastdeploy/model_executor/layers/sample/logprobs.py [hunk 11]
```diff
@@ -162,21 +163,22 @@ def build_output_logprobs(
...
 
-    real_bsz = share_inputs["seq_lens_this_time"].shape[0]
+    # NOTE(huicongyao) real_bsz is passed from _postprocess, remove this in future
+    max_occupied_slots = share_inputs["seq_lens_this_time"].shape[0]
 
     if is_naive:
         # NAIVE mode: one token per request, logits are already correct
         output_logits = logits
-        token_ids = share_inputs["accept_tokens"][:real_bsz, 0]
+        token_ids = share_inputs["accept_tokens"][:max_occupied_slots, 0]
```
[the remark this answers is not in the document]
Comment (huicongyao) [reply fastdeploy/model_executor/layers/sample/logprobs.py:172]: 这里用share_inputs["seq_lens_this_time"]的shape切分不会有问题

File: custom_ops/gpu_ops/speculate_decoding/build_sampling_params_logprob.cu [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+    start_offset += token_num_per_batch[i];
+  }
+
```
[the remark this answers is not in the document]
Comment (huicongyao) [reply custom_ops/gpu_ops/speculate_decoding/build_sampling_params_logprob.cu:36]: profiler实测128 bsz算子耗时约2us，没有性能问题

File: custom_ops/gpu_ops/speculate_decoding/build_sampling_params_logprob.cu [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+#define PD_BUILD_STATIC_OP(name) PD_BUILD_OP(static_op_##name)
+#endif
+
+template <typename T>
+__global__ void BuildSamplingParamLogProbKernel(
+    T* output_params,
+    const T* input_params,
+    const int32_t* token_num_per_batch,
+    const int64_t token_num_output_cpu,
+    int64_t real_bsz) {
```
[the remark this answers is not in the document]
Comment (huicongyao) [reply custom_ops/gpu_ops/speculate_decoding/build_sampling_params_logprob.cu]: 后续更新代码会先去掉

**Review by yongqiangma — APPROVED** [review]

**Review by freeliuzc — APPROVED** [review]

**Review by heavengate — APPROVED** [review]

**Review by xyxinyang — APPROVED** [review]
